### ISSUE TITLE
Docker Mount sysfs sub-directory RW without --privileged

### ISSUE THREAD

**Issue by wshen0123** [docker/docker#22825]
Hi, I have a app that need RW access to `/sys/class/uio/uio0/device/config` file. I can run docker container with `--privileged` option so the entire /sys/ is mounted RW but this is not secure since I only want that container to access uio0, not uio1.

I tried to` -v /sys/class/uio/uio0/:/sys/class/uio/uio0/` but in the container, when my app tried to open that config file with RW access, it said `config: Read-only file system`. Is this an issue with union fs that parent /sys/ is RO then it shadows sub-directories also as RO?

Thanks.

```
[root@ba83ec3d9d4c device]# cat /proc/mounts | grep sysfs
sysfs /sys sysfs ro,nosuid,nodev,noexec,relatime 0 0
sysfs /sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0/uio/uio0 sysfs rw,nosuid,nodev,noexec,relatime 0 0

[root@ba83ec3d9d4c device]# ll /sys/class/uio/uio0/device/config
-rw-r--r-- 1 root root 4096 May 19 07:10 /sys/class/uio/uio0/device/config
```
Note: /sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0/uio/uio0 is the actual directory pointed by symlink /sys/class/uio/uio0.

**Comment by justincormack** [comment]
You need to use the actual device or directory, not the one with the symlink in. A bind mount with `-v` is rw by default so you should be ok, but if you mount the symlink the actual device files are still in the `ro` sysfs mount.

**Comment by thaJeztah** [comment]
@justincormack would using `--device` work in this case? also see https://github.com/docker/docker/pull/22272

**Comment by wshen0123** [comment]
@justincormack thanks for the reply. Unfortunately when I specify `--volume=/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0:/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0` docker won't let me start the container
`docker: Error response from daemon: Invalid bind mount spec "/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0:/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0": Invalid volume specification: '/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0:/sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0'`

@thaJeztah using `--device` raises the same error as above.

**Comment by thaJeztah** [comment]
Invalid bind mount / volume specification is probably because the path you're mounting contains colons, so docker is trying to parse anything after the colon as volume-options. You can try escaping those

**Comment by justincormack** [comment]
Yes the colons will be the issue. I am not sure we support any form of escaping for them though, I will check the code.

**Comment by justincormack** [comment]
@thaJeztah no device won't work, it does not give access to these sysfs files which allow other changes and access.

**Comment by justincormack** [comment]
It is a bit roundabout but you could try on the host doing `mount --bind /sys/devices/pci0000:80/0000:80:01.0/0000:81:00.0 /tmp/mydev` and then doing `docker run -v /tmp/mydev:/mydev ...` and accessing it as `/mydev` - that should see if there are any issues apart from the issue with bind mount names.

**Comment by guhcampos** [comment]
I have recently stumbled into this exact same issue while working with hardware thar expose some functionality through its BAR in /sys, writing values to the BAR works as a way to send management commands to such device.

I attempted the long --mount, but it still mounts the directory read-only, this is how it looks:

`--mount type=bind,source=/sys/bus/pci/devices/0000:5a:00.0,target=/sys/bus/pci/devices/0000:5a:00.0`

Looks like there is no way to force the mount to be RW through this syntax. I can add `readonly` as an option, but not the opposite. I have even attempted to append `readonly=false` which docker happily accepted as valid input, but I still get Permission Denied when trying to write.

**Comment by amitinfo2k** [comment]
Facing the similar issue with :
Docker : v18.06.1-ce
Kubernetes : v1.13.1
---------------------------------------------------------------------------------
EAL: Cannot open /sys/class/uio/uio0/device/config: Read-only file system
---------------------------------------------------------------------------------
Any solution to this ?

**Comment by Nottt** [comment]
any news on this?

**Comment by git-developer** [comment]
A suggestion from the [Docker forum](https://forums.docker.com/t/unable-to-mount-sys-inside-the-container-with-rw-access-in-unprivileged-mode/97043/2) is to replace [runc](https://github.com/opencontainers/runc) by a different runtime (e.g. [sysbox](https://github.com/nestybox/sysbox)), which would mean a significant change to the host configuration. Are there any other workarounds known?

**Comment by yoursunny** [comment]
Docker version 20.10.17, build 100c701

It seems that `--mount type=bind,source=/sys,target=/sys` no longer mounts sysfs in RW mode.
Even if `mount` command inside the container shows `/sys` to be RW, opening `/sys/class/uio/uio0/device/config` for writing fails with EPERM.

The same setup worked in previous version of Docker (unsure which version, but it's about two years ago).

**Comment by jsofri** [comment]
same here - using docker 20.10.7.  
container's process writing to a file under `/sys/class/hwmon`, and succeeds only with `--privileged`.  
I tried running unprivileged with all capabilities (`--cap-add=`) but failed with ` Read-only file system`

**Comment by allfro** [comment]
Unfortunately, this is part of the default apparmor profile. You can try adding `--security-opt apparmor:unconfined` to see if it works.
